src/app/clusters/scenes/SceneTableImpl.cpp

```diff
@@ -0,0 +1,680 @@
+{
+    FabricSceneData fabric(fabric_index);
+    SceneTableData scene(fabric_index, scened_idx);
+
+    ReturnErrorOnFailure(scene.Delete(mStorage));
+
+    if (fabric.scene_count > 0)
+    {
+        fabric.scene_count--;
+        ReturnErrorOnFailure(fabric.Save(mStorage));
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+/// @brief Register a handler in the handler list
+/// @param handler Cluster specific handler for extension field sets interaction
+/// @return CHIP_NO_ERROR if handler was registered, CHIP_ERROR_NO_MEMORY if the handler list is full
+CHIP_ERROR DefaultSceneTableImpl::RegisterHandler(SceneHandler * handler)
+{
+    CHIP_ERROR err             = CHIP_ERROR_NO_MEMORY;
+    uint8_t idPosition         = kInvalidPosition;
+    uint8_t fisrtEmptyPosition = kInvalidPosition;
+
+    for (uint8_t i = 0; i < kMaxSceneHandlers; i++)
+    {
+        if (this->mHandlers[i] == handler)
+        {
+            idPosition = i;
+            break;
+        }
+        if (this->mHandlers[i] == nullptr && fisrtEmptyPosition == kInvalidPosition)
+        {
+            fisrtEmptyPosition = i;
+        }
+    }
+
+    // if found, insert at found position, otherwise at first free possition, otherwise return error
+    if (idPosition < kMaxSceneHandlers)
+    {
+        this->mHandlers[idPosition] = handler;
+        err                         = CHIP_NO_ERROR;
+    }
+    else if (fisrtEmptyPosition < kMaxSceneHandlers)
+    {
+        this->mHandlers[fisrtEmptyPosition] = handler;
+        this->mNumHandlers++;
+        err = CHIP_NO_ERROR;
+    }
+
+    return err;
+}
+
+CHIP_ERROR DefaultSceneTableImpl::UnregisterHandler(SceneHandler * handler)
+{
+    uint8_t position = kInvalidPosition;
+
+    // Verify list is populated and handler is not null
+    VerifyOrReturnValue(!this->HandlerListEmpty() && !(handler == nullptr), CHIP_NO_ERROR);
+
+    // Finds the position of the Handler to unregister
+    for (uint8_t i = 0; i < this->mNumHandlers; i++)
+    {
+        if (this->mHandlers[i] == handler)
+        {
+            position = i;
+            break;
+        }
+    }
+
+    // Verify Handler was found
+    VerifyOrReturnValue(position < kMaxSceneHandlers, CHIP_NO_ERROR);
+
+    uint8_t nextPos = static_cast<uint8_t>(position + 1);
+    uint8_t moveNum = static_cast<uint8_t>(kMaxSceneHandlers - nextPos);
+
+    // TODO: Implement general array management methods
+    // Compress array after removal
+    memmove(&this->mHandlers[position], &this->mHandlers[nextPos], sizeof(this->mHandlers[position]) * moveNum);
+
+    this->mNumHandlers--;
+    // Clear last occupied position
+    this->mHandlers[mNumHandlers] = nullptr;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DefaultSceneTableImpl::UnregisterAllHandler()
+{
+    for (uint8_t i = 0; i < this->mNumHandlers; i++)
+    {
+        ReturnErrorOnFailure(this->UnregisterHandler(this->mHandlers[0]));
```

Comment:
This loop is broken.  It will only remove half the handlers.  Is this tested?

Reply:
No longer relevant because we are using a linked list now, but for future reference: say we start this loop when mNumHandlers == 4.

* We enter the loop for i == 0, remove the first handler in the list.  mNumHandlers becomes 3, i becomes 1.
* We enter the loop for i == 1, remove the first handler in the list.  mNumHandlers becomes 2, i becomes 2.
* We exit the loop, since `i < mNumHandlers` tests false.

We used to have 4 handlers, we only removed two of them.

If there was a test for this and this test passed, then the test is broken and should be fixed.

Reply:
The test was broken because it executed both the UnregisterAllHandlers() and manually emptied all the table from 0 to kMaxSceneHandlers:
`    NL_TEST_ASSERT(aSuite, CHIP_NO_ERROR == sceneTable->UnregisterAllHandlers());
    for (uint8_t i = 0; i < scenes::kMaxSceneHandlers; i++)
    {
        NL_TEST_ASSERT(aSuite, CHIP_NO_ERROR == sceneTable->UnregisterHandler(&tmpHandler[i]));
    }`
    The test now only uses UnregisterAllHandlers and checks using the IsEmpty() method.